Multiple printings

BruebachL opened the pull request:
## Related Ticket(s)
- Fixes #IssueNumber

## Short roundup of the initial problem


## What will change with this Pull Request?
- this
- and this

## Screenshots
<!-- simply drag & drop image files directly into this description! -->

Commit:
Refactor CardInfo Widgets to reside in their appropriate folder and to have a clearer naming structure.

LunyaticCat committed:
canibalized zach work and added it to the decklist builder

Commit:
Change getPreferredPrintingForCard to getPreferredSetForCard to reflect refactor.

LunyaticCat committed:
Added Zach's work on storing printing information in the DeckList (#1)

* Change CardInfo's PixmapCacheKey to be the UUID of the preferred set after database loading has finished. Otherwise, and if no UUID of a preferred set is available, default to the card name.

* Refactor CardDatabase *db global variable to singleton CardDatabaseManager.

This commit refactors the global variable CardDatabase *db into a singleton encapsulated by the DatabaseManager class, accessible via DatabaseManager::getInstance(). This change centralizes access to the database instance, improving code modularity and encapsulation, resolving dependencies on main.h for code that requires access to the database instance.

- Added DatabaseManager class with getInstance() method returning a pointer to the singleton CardDatabase.
- Removed global db variable and updated references across the codebase.
 - Thread-safe static initialization for the singleton.

Impact: This refactor should have no functional impact on the application, as it maintains the same interface for accessing the CardDatabase instance. However, the codebase now benefits from improved encapsulation, lifetime management, and thread-safety.

* fixed db issue an renamed sets to set in picture loader

* canibalized zach work and added it to the decklist builder

---------

Co-authored-by: Lukas Brübach <Bruebach.Lukas@bdosecurity.de>

Commit:
Reintroduce some changes lost in the merge.

Commit:
Introduce UUID attribute to abstract_card_item, card_item, deck_view_card, server_card and serverinfo_card.

Commit:
Have various game events respect the new UUID attribute on instantiation.

Commit:
Correct some calls to default to preferred printing.

Commit:
DeckList now tries to assign reasonable defaults for UUID and collectorNumber if none are found in loaded DeckLists.
Rename overloaded DeckListModel findChild() function to findCardChildByNameAndUUID() for clarity.

Commit:
Properly update and set the DeckEditor's CardFrame to fetch by name and UUID if a card was selected from the decklist.

Commit:
Mainboard/Sideboard swaps should respect the UUID from the old zone instead of just blindly adding preferredPrinting.

Commit:
If the card info is null, there's no point in trying to look for the sets.

Commit:
Don't define methods twice.

Commit:
Convenience method to fetch a specific CardInfoPerSet instance for a cardName and a UUID.

Commit:
Check if the uuid starts with card_ when comparing.

Commit:
Address pull request comments (nullptr checks and additional comments, mostly.)

Commit:
Reformat code so the linter will stop yelling at me.

Commit:
DeckList no longer pre-populates uuids.

Commit:
Update Event_MoveCard to include the card UUID.

Commit:
Update Player::MoveCard to include the card UUID.

Commit:
Set the uuid when we set the cardName, in terms of hidden zones.

Commit:
[TEST/RevertMe] Set the uuid everywhere to test.

Commit:
Don't inline setUUID and mimic setName for AbstractCardItem.

Commit:
Revert blindly setting uuid for testing.

ZeldaZach commented on cockatrice/src/game/cards/abstract_card_item.cpp:
```diff
@@ -185,6 +188,19 @@ void AbstractCardItem::setName(const QString &_name)
+void AbstractCardItem::setUUID(const QString &_uuid)
+{
+    if (uuid == _uuid)
+        return;
```
brackets

ZeldaZach commented on cockatrice/src/game/cards/abstract_card_item.cpp:
```diff
@@ -185,6 +188,19 @@ void AbstractCardItem::setName(const QString &_name)
+
+    emit deleteCardInfoPopup(name);
+    if (info)
+        disconnect(info.data(), nullptr, this, nullptr);
```
brackets

ZeldaZach commented on common/decklist.cpp:
```diff
@@ -153,6 +155,18 @@ AbstractDecklistNode *InnerDecklistNode::findChild(const QString &_name)
+{
+    for (int i = 0; i < size(); i++) {
+        if (at(i) != nullptr) {
+            if (at(i)->getName() == _name && at(i)->getCardUuid() == _uuid) {
```
can combine into one if statement

Commit:
Address PR comments (AbstractCardItem).

Commit:
Combine if-statement.

Commit:
Re-order uuid to visually align with its field number.

Commit:
Remove unnecessary new uuid field from event_move_card.

BruebachL committed:
Refactor Card Info Widgets (#5176)

* Refactor CardInfo Widgets to reside in their appropriate folder and to have a clearer naming structure.

* Add optional HoverToZoom functionality to CardInfoPictureWidget (default: disabled) and utility class to display text over a CardInfoPictureWidget.

* Patch CardInfoWidgets to use the new CardDatabaseManager.

* Add HoverToZoom to CardInfoPictureWithTextOverlayWidget

* Refactors and new signals for CardInfoPictureWidgets.

* Address pull request comments (nullptr checks and additional comments, mostly.)

* Reformat code so the linter will stop yelling at me.

* Linting.

* Fix the build.

* Fix warnings.

* Formatting, const qualifiers.

* Sensibly call the base class's (QWidget) paint event.

* Address PR comments (card picture).

* QT Version check because enterEvent signature changed.

* Linting.

---------

Co-authored-by: Lukas Brübach <Bruebach.Lukas@bdosecurity.de>

BruebachL committed:
Add Utility Layouts and corresponding Widgets (#5177)

* Add FlowWidget class with flexible layout and scroll handling

- Implemented FlowWidget class to organize widgets in a flow layout with scrollable options.
- Integrated QScrollArea to handle overflow with configurable horizontal and vertical scroll policies.
- Incorporated dynamic layout selection (FlowLayout, HorizontalFlowLayout, VerticalFlowLayout) based on scroll policy.

* Add OverlapWidget and OverlapLayout for managing overlapping child widgets

- Implemented the OverlapWidget class to manage child widgets in an overlapping manner, supporting configurable overlap percentage, maximum columns, maximum rows, and layout direction.
- Introduced the OverlapLayout class, which arranges widgets with overlapping positions, allowing flexible stacking based on specified parameters.

* Add OverlapControlWidget.

* Delete FlowLayout items later to allow them to finish their event loop.

* Allow OverlapWidgets to adjust their rows/columns on resize.

* Clamp vertical FlowLayout to any available parent scrollAreas.

* Implement margins and spacing for FlowLayouts.

* Adjust/revert some things.

* Address pull request comments (nullptr checks and additional comments, mostly.)

* Reformat code so the linter will stop yelling at me.

* Remove undefined methods from FlowLayouts.

* Fix the build.

* Revert FlowLayout::takeAt to index check.

* Commits will continue until linter morale improves.

* Fix various warnings in FlowLayout.

* Fix various warnings in FlowLayout.h.

* Fix various warnings in the FlowLayout classes.

* Fix [[nodiscard]] warning.

* Fix more warnings.

* Final round of yellow squiggle fixing.

* Linter formatting.

* Refactor column/row calculation to be more readable.

* Code style.

* Address PR comments.

* Combine if-statements.

* Replace std::math functions with Qt equivalents.

* Fix non-consts and QtMath.

---------

Co-authored-by: Lukas Brübach <Bruebach.Lukas@bdosecurity.de>

tooomm committed:
Readme: Remove last Gitter hint (#5178)

Commit:
Remove unused imports.

RickyRister committed:
don't open deck in new tab if current tab is blank (#5169)

Commit:
Include cardName in the PixmapCacheKey in order to not break double-faced cards.

Commit:
Refactor setCode to cardUUID and introduce new cardSetShortName field.

ZeldaZach committed:
Override

Commit:
Refactor UUID to be providerId and change QString comparisons with empty string to isEmpty().

Commit:
Update translations.

Commit:
Change parent to be the first argument.

Commit:
Pull Parent argument up for CardItem.

Commit:
Pull Parent argument up for CardItem.

Commit:
Linter.
